framework/rendering/subpass.h

```diff
@@ -40,217 +50,442 @@ struct alignas(16) Light
+	std::copy(lighting_state.point_lights.begin(), lighting_state.point_lights.end(), light_info.point_lights);
+	std::copy(lighting_state.spot_lights.begin(), lighting_state.spot_lights.end(), light_info.spot_lights);
 
-		auto &render_frame          = get_render_context().get_active_frame();
-		lighting_state.light_buffer = render_frame.allocate_buffer(VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT, sizeof(T));
-		lighting_state.light_buffer.update(light_info);
-	}
+	auto &render_frame          = render_context.get_active_frame();
+	lighting_state.light_buffer = render_frame.allocate_buffer(vk::BufferUsageFlagBits::eUniformBuffer, sizeof(T));
+	lighting_state.light_buffer.update(light_info);
+}
 
-  protected:
-	RenderContext &render_context;
+template <vkb::BindingType bindingType>
+inline const std::vector<uint32_t> &Subpass<bindingType>::get_color_resolve_attachments() const
+{
+	return color_resolve_attachments;
+}
 
-	VkSampleCountFlagBits sample_count{VK_SAMPLE_COUNT_1_BIT};
+template <vkb::BindingType bindingType>
+inline const std::string &Subpass<bindingType>::get_debug_name() const
+{
+	return debug_name;
+}
 
-	// A map of shader resource names and the mode of constant data
-	std::unordered_map<std::string, ShaderResourceMode> resource_mode_map;
+template <vkb::BindingType bindingType>
+inline const uint32_t &Subpass<bindingType>::get_depth_stencil_resolve_attachment() const
+{
+	return depth_stencil_resolve_attachment;
+}
 
-	/// The structure containing all the requested render-ready lights for the scene
-	LightingState lighting_state{};
+template <vkb::BindingType bindingType>
+inline typename Subpass<bindingType>::ResolveModeFlagBitsType Subpass<bindingType>::get_depth_stencil_resolve_mode() const
+{
+	if constexpr (bindingType == vkb::BindingType::Cpp)
+	{
+		return depth_stencil_resolve_mode;
+	}
+	else
+	{
+		return static_cast<VkResolveModeFlagBits>(depth_stencil_resolve_mode);
+	}
+}
 
-  private:
-	std::string debug_name{};
+template <vkb::BindingType bindingType>
+inline typename Subpass<bindingType>::DepthStencilStateType &Subpass<bindingType>::get_depth_stencil_state()
+{
+	if constexpr (bindingType == vkb::BindingType::Cpp)
+	{
+		return depth_stencil_state;
+	}
+	else
+	{
+		return reinterpret_cast<vkb::DepthStencilState &>(depth_stencil_state);
+	}
+}
 
-	ShaderSource vertex_shader;
+template <vkb::BindingType bindingType>
+inline const bool &Subpass<bindingType>::get_disable_depth_stencil_attachment() const
+{
+	return disable_depth_stencil_attachment;
+}
 
-	ShaderSource fragment_shader;
+template <vkb::BindingType bindingType>
+inline const ShaderSource &Subpass<bindingType>::get_fragment_shader() const
+{
+	return fragment_shader;
+}
 
-	DepthStencilState depth_stencil_state{};
+template <vkb::BindingType bindingType>
+inline void Subpass<bindingType>::set_color_resolve_attachments(std::vector<uint32_t> color_resolve)
+{
+	color_resolve_attachments = color_resolve;
```

Comment:
copy a vector on each call.  maybe move is better?

Reply:
Good catch!
Will make it a `std::vector<uint32_t> const &`